crates/matrix-sdk/src/event_cache/room/threads/mod.rs

```diff
@@ -0,0 +1,404 @@
+            .map_err(|err| {
+                EventCacheThreadError::SummaryDeserialization(format!(
+                    "unable to deserialize an event id in a thread summary: {err}",
+                ))
+            })?
+            .ok_or_else(|| {
+                EventCacheThreadError::SummaryDeserialization(
+                    "missing event id in the latest thread event of a thread summary".to_owned(),
+                )
+            })?;
+
+        Ok(Some(ThreadSummary {
+            count,
+            latest_event,
+            current_user_participated: bundled_thread.current_user_participated,
+            maybe_outdated: true,
+        }))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::ops::Not as _;
+
+    use assert_matches::assert_matches;
+    use matrix_sdk_test::{event_factory::EventFactory, ALICE};
+    use ruma::{
+        event_id,
+        events::{relation::BundledThread, BundledMessageLikeRelations},
+        room_id, uint,
+    };
+
+    use super::{RoomThreads, ThreadSummary};
+    use crate::event_cache::room::threads::EventCacheThreadError;
+
+    #[test]
+    fn test_extract_thread_summary() {
+        let room_id = room_id!("!room:example.org");
+        let f = EventFactory::new().sender(&ALICE).room(room_id);
+
+        let latest_event_id = event_id!("$latest");
+        let latest_event = f.text_msg("hello to you too!").event_id(latest_event_id).into_raw();
+        let bundled_thread = BundledThread::new(latest_event, uint!(3), false);
+        let mut relations = BundledMessageLikeRelations::new();
+        relations.thread = Some(Box::new(bundled_thread));
+
+        let root = f.text_msg("hello world!").bundled_relations(relations).into_raw();
+
+        let extracted_summary = ThreadSummary::extract_from_bundled(&root)
+            .expect("extracting works")
+            .expect("summary found");
+
+        assert_eq!(extracted_summary.count, 3);
+        assert!(extracted_summary.current_user_participated.not());
+        assert_eq!(extracted_summary.latest_event, latest_event_id);
+
+        // Summaries are assumed to be outdated when extracted the first time.
+        assert!(extracted_summary.maybe_outdated);
+    }
+
+    #[test]
+    fn test_extract_no_thread_summary() {
+        let room_id = room_id!("!room:example.org");
+        let f = EventFactory::new().sender(&ALICE).room(room_id);
+
+        let root = f.text_msg("hello world!").into_raw();
+
+        // If there's no thread summary, it can be extracted without trouble.
+        let extracted_summary =
+            ThreadSummary::extract_from_bundled(&root).expect("extracting works");
+        assert_matches!(extracted_summary, None);
+
+        // If there's a bundled relation, but it's not a thread, we can still extract
+        // without failing.
+        let mut relations = BundledMessageLikeRelations::new();
+        relations.replace = Some(Box::new(f.text_msg("hello wrold!").into_raw()));
```

Comment:
hah, it was supposed to be the replacement event for a message that had a typo in it, but of course the typos bot is unhappy with such a test setup krkrkr